File: SDK/c/examples/subscribe_status.c
/* subscribe_status.c - Demonstrates real-time status subscription with a
 * callback, mirroring examples/subscribe_status.py. */

#if !defined(_WIN32)
#  ifndef _POSIX_C_SOURCE
#    define _POSIX_C_SOURCE 200809L
#  endif
#endif

#include "csm_tcp_router_client.h"

#include <stdio.h>
#include <stdlib.h>

#if defined(_WIN32)
#  include <windows.h>
static void sleep_ms(unsigned int ms){ Sleep(ms); }
#else
#  include <time.h>
static void sleep_ms(unsigned int ms){
    struct timespec ts; ts.tv_sec=ms/1000; ts.tv_nsec=(long)(ms%1000)*1000000L;
    nanosleep(&ts, NULL);
}
#endif

#define HOST "localhost"
#define PORT 30007

static void on_status(const csm_status_notification_t *n, void *ud) {
    (void)ud;
    printf("[%s @ %s] %s\n", n->status_name, n->module_name, n->data);
}

int main(int argc, char **argv) {
    const char *status_name = (argc > 1) ? argv[1] : "Status";
    const char *module_name = (argc > 2) ? argv[2] : "DAQmx";

    csm_client_t *c = csm_client_create();
    if (!c) return 1;

    csm_result_t r = csm_client_connect(c, HOST, PORT, 5000);
    if (r != CSM_OK) {
        fprintf(stderr, "Connection failed: %s\n", csm_result_str(r));
        csm_client_destroy(c);
        return 1;
    }

    r = csm_client_subscribe_status(c, status_name, module_name,
                                    on_status, NULL, 5000);
    if (r != CSM_OK) {
        fprintf(stderr, "Subscribe failed: %s\n", csm_result_str(r));
        csm_client_disconnect(c);
        csm_client_destroy(c);
        return 1;
    }

    printf("Subscribed to %s@%s. Listening for 30s ...\n",
           status_name, module_name);
    sleep_ms(30000);

    csm_client_unsubscribe_status(c, status_name, module_name, 5000);
    csm_client_disconnect(c);
    csm_client_destroy(c);
    return 0;
}

File: SDK/python/examples/subscribe_status.py
"""Status subscription example for csm-tcp-router-client.

Prerequisites
-------------
A running CSM-TCP-Router server that has a CSM module publishing a status.
The reference app (``CSM-TCP-Router(Server).vi``) exposes an ``AI`` module
that continuously broadcasts a ``Status`` status.

Install the SDK::

    pip install csm-tcp-router-client

Run this example::

    python subscribe_status.py
"""

import signal
import threading
import time

from csm_tcp_router_client import ConnectionError, ServerError, StatusNotification, TcpRouterClient

HOST = "localhost"
PORT = 30007

# Module and status name to subscribe to (adjust to match your server)
MODULE_NAME = "AI"
STATUS_NAME = "Status"

# Global stop flag
_stop = threading.Event()


def on_status(notification: StatusNotification) -> None:
    """Callback invoked on every status broadcast from the server."""
    print(
        f"[{time.strftime('%H:%M:%S')}] "
        f"{notification.status_name} @ {notification.module_name} "
        f"→ {notification.data}"
    )


def main() -> None:
    # Allow Ctrl-C to exit cleanly
    signal.signal(signal.SIGINT, lambda *_: _stop.set())

    with TcpRouterClient() as client:
        try:
            client.connect(HOST, PORT)
        except ConnectionError as exc:
            print(f"Connection failed: {exc}")
            return

        print(f"Connected to {HOST}:{PORT}")

        # Subscribe to status broadcasts
        try:
            client.subscribe_status(STATUS_NAME, MODULE_NAME, callback=on_status)
            print(
                f"Subscribed to '{STATUS_NAME}' from module '{MODULE_NAME}'. "
                "Press Ctrl-C to exit.\n"
            )
        except ServerError as exc:
            print(f"Subscription failed: {exc}")
            return

        # Keep running until Ctrl-C
        while not _stop.is_set():
            # You can also poll client.status_queue here if you prefer
            # notification = client.status_queue.get(timeout=1.0)
            time.sleep(0.1)

        # Unsubscribe cleanly before disconnecting
        try:
            client.unsubscribe_status(STATUS_NAME, MODULE_NAME)
            print("\nUnsubscribed.")
        except Exception:
            pass

    print("Disconnected.")


if __name__ == "__main__":
    main()
